### ISSUE TITLE
for_where doesn't catch nested ifs with pattern-matching

### ISSUE THREAD

**Issue by tomquist** [realm/SwiftLint#2053]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Bug Report

I think the change done in #1978 is based on wrong assumptions and should be undone. The example outlined in #1968 can be expressed as follows:

```swift
enum Value {
    case valueA, valueB(a: Int)
}

func firstIndexOfValueB(in array: [Value]) -> Int? {
    for case let (index, .valueB) in array.enumerated() {
        return index
    }
    return nil
}
```

Maybe the name of the rule is misleading because there is no `where` keyword in the for-loop. There are two possible solutions:
1. Undo the change in #1978 and change the description of the rule to "`where` clauses or pattern-matching are preferred over a single `if` inside a `for`.". The name should probably also be changed but the identifier has to stay the same for compatibility reasons.
2. Don't undo the change and create another rule to catch only the pattern-matching case

**Comment by marcelofabri** [comment]
@tomquist Can you take a look on #1977? I think this is a duplicate of that rule request.

**Comment by tomquist** [comment]
@marcelofabri I agree that these have similarities but #1977 explicitly talks about unwrapping of optionals with a `guard` statement which should be autocorrected but I'm talking about pattern matching in general. Maybe it also makes sense to merge #2053 and #1977.
